compiler/rustc_ty_utils/src/ty.rs

```diff
@@ -275,8 +278,14 @@ impl<'tcx> TypeVisitor<TyCtxt<'tcx>> for ImplTraitInTraitFinder<'_, 'tcx> {
             // We have entered some binders as we've walked into the
             // bounds of the RPITIT. Shift these binders back out when
             // constructing the top-level projection predicate.
-            let alias_ty = self.tcx.fold_regions(alias_ty, |re, _| {
+            let shifted_alias_ty = self.tcx.fold_regions(alias_ty, |re, depth| {
```

Comment:
When we turn `alias_ty` into `shifted_alias_ty`, we shift all the binders from `self.depth` to `ty::INNERMOST` so we have no escaping bound vars in the projection we collect at 285. 

So when we walk the type's bounds on line 309, we're no longer conceptually `alias_ty`'s substs are technically only now one binder deep, so shifting any nested impl traits would result in us shifting to a negative de bruijn index.